src/array/cpu/segment_reduce.h

```diff
@@ -117,61 +118,35 @@ void UpdateGradMinMax_hetero(HeteroGraphPtr graph,
                        const std::string& op,
                        const std::vector<NDArray>& list_feat,
                        const std::vector<NDArray>& list_idx,
-                       const std::vector<NDArray>& list_idx_ntypes,
+                       const std::vector<NDArray>& list_idx_types,
                        std::vector<NDArray>* list_out) {
-  if (op == "copy_lhs") {
-    std::vector<std::vector<dgl_id_t>> dst_src_ntids(graph->NumVertexTypes(),
+  if (op == "copy_lhs" || op == "copy_rhs") {
+    std::vector<std::vector<dgl_id_t>> src_dst_ntypes(graph->NumVertexTypes(),
     std::vector<dgl_id_t>());
+
     for (dgl_type_t etype = 0; etype < graph->NumEdgeTypes(); ++etype) {
       auto pair = graph->meta_graph()->FindEdge(etype);
-      const dgl_id_t dst_id = pair.first;  // graph is reversed
-      const dgl_id_t src_id = pair.second;
-      dst_src_ntids[dst_id].push_back(src_id);  // can have duplicates. Use Hashtable to optimize.
-    }
-    std::vector<bool> updated(graph->NumVertexTypes());
-    for (int dst_id = 0; dst_id < dst_src_ntids.size(); ++dst_id) {
-      std::fill(updated.begin(), updated.end(), false);
-      for (int j = 0; j < dst_src_ntids[dst_id].size(); ++j) {
-        int src_id = dst_src_ntids[dst_id][j];
-        if (updated[src_id]) continue;
-        const DType* feat_data = list_feat[dst_id].Ptr<DType>();
-        const IdType* idx_data = list_idx[dst_id].Ptr<IdType>();
-        const IdType* idx_ntype_data = list_idx_ntypes[dst_id].Ptr<IdType>();
-        DType* out_data = (*list_out)[src_id].Ptr<DType>();
-        int dim = 1;
-        for (int i = 1; i < (*list_out)[src_id]->ndim; ++i)
-          dim *= (*list_out)[src_id]->shape[i];
-        int n = list_feat[dst_id]->shape[0];
-#pragma omp parallel for
-        for (int i = 0; i < n; ++i) {
-          for (int k = 0; k < dim; ++k) {
-            if (src_id == idx_ntype_data[i * dim + k]) {
-              const int write_row = idx_data[i * dim + k];
-#pragma omp atomic
-              out_data[write_row * dim + k] += feat_data[i * dim + k];  // feat = dZ
-            }
-          }
-        }
-        updated[src_id] = true;
-      }
-    }
-  } else if (op == "copy_rhs") {
-    for (dgl_type_t etid = 0; etid < graph->NumEdgeTypes(); ++etid) {
-      auto pair = graph->meta_graph()->FindEdge(etid);
-      const dgl_id_t dst_id = pair.first;  // graph is reversed
-      const dgl_id_t src_id = pair.second;
-      const DType* feat_data = list_feat[dst_id].Ptr<DType>();
-      const IdType* idx_data = list_idx[dst_id].Ptr<IdType>();
-      const IdType* idx_ntype_data = list_idx_ntypes[dst_id].Ptr<IdType>();
-      DType* out_data = (*list_out)[etid].Ptr<DType>();
+      const dgl_id_t dst_ntype = pair.first;  // graph is reversed
+      const dgl_id_t src_ntype = pair.second;
+      auto same_src_dst_ntype = std::find(begin(src_dst_ntypes[dst_ntype]),
+        end(src_dst_ntypes[dst_ntype]), src_ntype);
```

Comment:
Where is `begin` and `end` defined?  Somewhere must `using`'d `std::begin` and `std::end`, which is bad since `begin` and `end` alone can very well be used as variable names.  I prefer explicitly use `std::begin` and `std::end` (and find and remove the `using std::begin` and `using std::end` statements).

Reply:
Agreed. Changed in CPU and GPU version.